pkg/apis/apps/types.go

```diff
@@ -157,6 +184,13 @@ type StatefulSetSpec struct {
 	// consists of all revisions not represented by a currently applied
 	// StatefulSetSpec version. The default value is 10.
 	RevisionHistoryLimit *int32
+
+	// PersistentVolumeClaimDeletePolicy indicates the policy that will
+	// determine when volumes from the VolumeClaimTemplates will be deleted when
+	// the controlling StatefulSet is deleted or scaled down. This requires
+	// the StatefulSetAutoDeletePVC feature gate.
+	// +optional
+	PersistentVolumeClaimDeletePolicy *PersistentVolumeClaimDeletePolicyType
```

Comment:
It doesn't look like a pointer adds any value. Unset can be identified from empty string.

Disregard the following if it was already discussed in the KEP.

Have you thought of future policies for PVCs? If so, would it make sense to do something like:

```golang
type PersistentVolumeClaimPolicy struct {
  Delete PersistentVolumeClaimDeletePolicyType
}
```

Reply:
I was going off of https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version step 2 where it says that optional fields should be pointers.

LMK if I misinterpreted.

As to the need for future policies, there are no plans, so it seems premature. OTOH, if there ever is a future one it would be nice. So I'll let you advise me on it.

Reply:
This field is not really optional, as it always has a value, "Retain" in this case. Also, empty string is "nil value" in this case https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

A pointer would be advised if empty string was a valid value different than the default.

> As to the need for future policies, there are no plans, so it seems premature. OTOH, if there ever is a future one it would be nice. So I'll let you advise me on it.

I think that falls more in the sig to decide.

Reply:
> Have you thought of future policies for PVCs?

Since this looks to be adding a new enum-like field, consider https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas

_For enumeration-like fields that expect to add new values in the future, such as reason fields, document that expectation clearly in the API field description in the first release the field is made available, and describe how clients should treat an unknown value. Clients should treat such sets of values as potentially open-ended._

If you don't explicitly declare this is open-ended and tell clients how they should treat unknown values, you won't be able to add new values in the future.

>  If so, would it make sense to do something like:
>  
>  ```
>  type PersistentVolumeClaimPolicy struct {
>    Delete PersistentVolumeClaimDeletePolicyType
>  }
>  ```

Many of our strategy/policy fields have associated parameters in other types (xref DeploymentStrategy and StatefulSetUpdateStrategy) which have a type and then a discriminated union of subfields with options per type

Reply:
Update regarding pointer: while omitempty doesn't render empty string in json/yaml, it does in proto. So we need the pointer :/

Reply:
Okay, makes sense. I'll update it to a struct (in which case the pointer issue goes away...)

Reply:
Updated to be a struct, PTAL.

Reply:
On second thought, `delete: Retain` doesn't read very well. We could try to find better names for the field/values, but I think it makes sense to split it:

```yaml
persistentVolumeClaimPolicy:
  onStatefulSetDeletion: Retain|Delete
  onScaleDown: Retain|Delete
```

cc @msau42

Reply:
This had been done.